specs/sharding/beacon-chain.md

```diff
@@ -587,36 +649,54 @@ def update_pending_shard_work(state: BeaconState, attestation: Attestation) -> N
             )
 ```
 
+
+#### `charge_builder`
+
+```python
+def charge_builder(state: BeaconState, index: BuilderIndex, fee: Gwei) -> None:
+    """
+    Decrease the builder balance at index ``index`` by ``fee``, with underflow check.
+    """
+    assert state.builder_balances[index] >= fee
+    state.builder_balances[index] -= fee
```

Comment:
I think a single balance for a builder can cause issues for shard proposers. 

take the following:
* `builder_balance == 100`
* `builder` creates `blob` with fee `10`
* `proposer` signs `blob`
* `builder` has `secret_blob` with fee `100`
* `builder` gets `secret_proposer` to sign `secret_blob`
* `secret_blob` and `blob` now cannot both be included. `secret_blob` is included first, thus invalidating honest `blob`

Reply:
Yes, the assumption is that there is some minimum balance of a builder that we'll need to enforce to avoid race conditions in accepting fees. But with arbitrary priority fees, that may not be so simple. Requiring the builder to have 64 times the fee worth of balance may work, but is inefficient. We'll decide on a better approach later.